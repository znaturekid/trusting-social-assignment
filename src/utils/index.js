export const formatDate = (dateStr) => {
    if (!dateStr) {
        return null;
    }
    return new Date(dateStr).toDateString();
}

export const getFirstImageLink = (multimedia) => {
    return (multimedia && multimedia[0] && `https://static01.nyt.com/${multimedia[0].url}`) || null;
}

export const requestJSON = async (link, options) => {
    return new Promise(async (resolve, reject) => {
        try {
            const response = await fetch(link, options);
            const data = response.json();
            resolve(data);
        } catch (e) {
            reject(e);
        }
    })
}