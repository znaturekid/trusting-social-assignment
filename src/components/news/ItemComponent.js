import React from 'react';
import styled from 'styled-components';
import { formatDate, getFirstImageLink } from '../../utils';
import { withRouter } from 'react-router';
import PropTypes from 'prop-types';
const Img = styled.div`
    @keyframes increase {
        0% {
            transform: scale(1);
        }
        100% {
            opacity: 1;
            transform: scale(1.1);
        } 
    }
    position: relative;
    background-image: ${ props => `url(${props.src})` || null };
    background-position: 50% 50%;
    background-size: contain;
    transition: all 0.4s ease-in-out;
    height: 240px;
    &:hover {
        background-size: cover;
        cursor: pointer;
    }
    & ~ div {
        margin-top: 0px;
    }
`
const Item = styled.div`
    box-shadow: 1px 1px 6px #404040;
    display: grid;
    background: rgba(226, 226, 226, 0.6);
    position: relative;
    overflow: hidden;
` 

const InfoWrapper = styled.div`
    display: grid; 
    grid-template-columns: 1fr 1fr;
    grid-template-rows: auto;
    align-content: flex-end;
    color: #171616;
    padding: 10px;
    cursor: pointer;
    margin-top: 30px;
`
const Snippet = styled.p`
    grid-column: 1 / span 2;
`

const Source = styled.span`
    font-size:12px;   
    position: absolute;
    left: 0;
    top: 10px;
    background: rgba(0,0,0,0.6);
    padding: 5px 10px;
    color: #fff;
`
const Date = styled.span`
    position: absolute;
    font-size: 12px;
    right: 0;
    top: 10px;
    background: rgba(0,0,0,0.6);
    padding: 5px 10px;
    color: #fff;
`


class ItemComponent extends React.Component {
    constructor(props) {
        super(props);
        this.viewDetail = this.viewDetail.bind(this)
    }
    viewDetail () {
        const { onClick, item, history } = this.props;
        onClick && item && onClick(item);
        history && history.push('/detail');
    }
    
    render () {
        const { item } = this.props;
        const img = getFirstImageLink(item.multimedia);
        return (
            <Item onClick={this.viewDetail}>
                {img && <Img src={img}/>}
                {item.pub_date && <Date>{formatDate(item.pub_date)}</Date>}
                {item.source && <Source>{item.source}</Source>}
                <InfoWrapper>
                    <Snippet>{item.snippet}</Snippet>
                </InfoWrapper>
            </Item>)
    }
}

ItemComponent.propTypes = {
    item: PropTypes.object.isRequired,
    onClick: PropTypes.func.isRequired
}

export default withRouter(ItemComponent)