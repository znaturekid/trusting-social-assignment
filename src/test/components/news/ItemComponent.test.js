import React from 'react';
import Enzyme from 'enzyme';
import { formatDate } from '../../../utils';
import Adapter from 'enzyme-adapter-react-16'
import ItemComponent from '../../../components/news/ItemComponent'
import 'jest-styled-components';
import { mountWrap } from '../../help';
Enzyme.configure({
    adapter: new Adapter()
});

function setup() {
    const props = {
        item: {
            source: 'source',
            snippet: 'snippet',
            pub_date: '2018-08-05T21:21:27+0000',
            multimedia: [{url: 'link.png'}]
        },
        onClick: jest.fn(),
    }

    const component = (<ItemComponent { ...props} />)

    return {
        props,
        component
    }
}

describe('components', () => {
    describe('Item =>', () => {
        it('should render item', () => {
            const { component, props} = setup()
            const list = mountWrap(component);
            expect(list.find('p').text()).toBe(props.item.snippet);
            expect(list.find('span').at(0).text()).toBe(formatDate(props.item.pub_date));
            expect(list.find('span').at(1).text()).toBe(props.item.source);
            list.simulate('click');
            expect(props.onClick).toBeCalledWith(props.item);
        })

        it('should call onClick when click on the item', () => {
            const { component, props} = setup()
            const list = mountWrap(component);
            list.simulate('click');
            expect(props.onClick).toBeCalledWith(props.item);
        })
    })
})