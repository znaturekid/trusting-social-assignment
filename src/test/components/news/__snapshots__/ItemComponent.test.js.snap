// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Item => should render item 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.span>
    source
  </styled.span>
  <styled.div>
    <styled.p>
      snippet
    </styled.p>
  </styled.div>
</styled.div>
`;
